Replace Channelable interface with a dedicated CRD

scothis opened the pull request:
- add ChannelSubscriptionSet CRD
- remove Channelable interface
- define that a ChannelSubscriptionSet is created for a Channel by the
  Provisioner and referenced in the ChannelStatus
- update definition of Subscribeable

Fixes knative/eventing#512

This PR conflicts with #4, but is authored to allow either to merge merged independently. The second to merge will need to be rebased.

scothis committed:
Replace Channelable interface with a dedicated CRD

- add ChannelSubscriptionSet CRD
- remove Channelable interface
- define that a ChannelSubscriptionSet is created for a Channel by the
  Provisioner and referenced in the ChannelStatus
- update definition of Subscribeable

Fixes knative/eventing#512

grantr commented on docs/spec/interfaces.md:
```diff
@@ -15,33 +15,15 @@ can direct to their configured destination.
-_Channelable_ resource; the field MAY refer back to this _Subscribable_ as a
+_status.subscriptions_ field (an _ObjectReference_). The resource
+referenced in the _status.subscriptions_ field MUST be a
+_ChannelSubscriptionSet_ resource; the field MAY refer back to this _Subscribable_ as a
```
I'm confused about the self-referencing aspect. Is _ChannelSubscriptionSet_ also _Subscribable_?

scothis replied:
Yea the self referencing bit is confusing, that's a holdover that I should have removed.

I also proposed that the Subscribable interface be removed in #4, but I'll update this PR to make more sense individually.

grantr commented on docs/spec/spec.md:
```diff
@@ -121,6 +122,31 @@ Subscription's call parameter._
+### group: eventing.internal.knative.dev/v1alpha1
+
+_A ChannelSubscriptionSet holds an aggregation of resolved subscriptions for a
+Channel._
```
A couple unknowns I'd like to see specified:
- What is responsible for creating this object? Subscription controller, channel provisioner, something else?
- When is it created? Channel creation time or subscription creation time?

scothis replied:
I thought I had a bit saying the ChannelSubscriptionSet should be created by the Provisioner as part of provisioning the channel, but I don't see it. Will add.

grantr commented on docs/spec/spec.md:
```diff
@@ -121,6 +122,31 @@ Subscription's call parameter._
+
+| Field         | Type                    | Description                                                           | Limitations                            |
+| ------------- | ----------------------- | --------------------------------------------------------------------- | -------------------------------------- |
+| subscribers\* | ChannelSubscriberSpec[] | Information about subscriptions used to implement message forwarding. | Filled out by Subscription Controller. |
```
To clarify what Subscription controller actually does:
1. Subscription controller reconciles a Subscription.
1. If the Subscription names a Channel in its `from`, Subscription controller gets that Channel then gets the ChannelSubscriptionSet in its Status.
1. Subscription controller looks for a subscriber in ChannelSubscriptionSet matching the Subscription being reconciled, and adds one if none is found. (There's an issue here around losing a reference to the original subscription, so Subscription updates don't work and duplicate subscriptions aren't possible, but that's an issue with the current spec also.)

Is this correct?

scothis replied:
I lifted this line from the current [Channelable definition](https://github.com/n3wscott/eventing/pull/5/files/19ebafb9283c2122724bbe441fbd1d0e32decfeb#diff-bd2ff29d22d19f9b6c62779fda3647d8L309), the only change was to make the field required.

What you listed is the intended behavior. I thought I had also stated this explicitly. I'll add a clarification.

grantr reviewed (commented):
Thanks @scothis, it's helpful to see a concrete design.

scothis committed:
Review feedback

- renamed ChannelSubscriptionSet to ResolvedSubscriptionSet
- renamed ChannelSubscriptionSpec to ResolvedSubscription
- added details about the Subscription reconciliation process

scothis committed:
missed one

scothis commented:
@grantr PTAL

n3wscott commented:
I think we will hold on this one.

scothis commented:
Will create a new PR based on all of the other movement.
